consensus/qbft/core/preprepare.go

```diff
@@ -0,0 +1,128 @@
+	logger := c.logger.New("state", c.state)
+
+	// If I'm the proposer and I have the same sequence with the proposal
+	if c.current.Sequence().Cmp(request.Proposal.Number()) == 0 && c.IsProposer() {
+		curView := c.currentView()
+		preprepare := message.NewPreprepare(curView.Sequence, curView.Round, request.Proposal)
+		preprepare.SetSource(c.Address())
+
+		// Sign payload
+		encodedPayload, err := preprepare.EncodePayloadForSigning()
+		if err != nil {
+			logger.Error("QBFT: Failed to encode payload of pre-prepare message", "msg", preprepare, "err", err)
+			return
+		}
+		signature, err := c.backend.Sign(encodedPayload)
+		if err != nil {
+			logger.Error("QBFT: Failed to sign pre-prepare message", "msg", preprepare, "err", err)
+			return
+		}
+		preprepare.SetSignature(signature)
+
+		// Justification
+		if request.RCMessages != nil {
+			preprepare.JustificationRoundChanges = make([]*message.SignedRoundChangePayload, 0)
+			for _, m := range request.RCMessages.Values() {
+				preprepare.JustificationRoundChanges = append(preprepare.JustificationRoundChanges, &m.(*message.RoundChange).SignedRoundChangePayload)
+				logger.Info("QBFT: Appending RC justification", "rc", m.(*message.RoundChange).SignedRoundChangePayload)
+			}
+			logger.Info("QBFT: On Pre-prepare", "rc justification", preprepare.JustificationRoundChanges)
+		}
+		if request.PrepareMessages != nil {
+			preprepare.JustificationPrepares = request.PrepareMessages
+			logger.Info("QBFT: On Pre-prepare", "prepare justification", preprepare.JustificationPrepares)
+		}
+
+		// RLP-encode message
+		payload, err := rlp.EncodeToBytes(&preprepare)
+		if err != nil {
+			logger.Error("QBFT: Failed to encode pre-prepare message", "msg", preprepare, "err", err)
+			return
+		}
+
+		logger.Info("QBFT: sendPreprepareMsg", "m", preprepare, "payload", hexutil.Encode(payload))
+		// Broadcast RLP-encoded message
+		if err = c.backend.Broadcast(c.valSet, preprepare.Code(), payload); err != nil {
+			logger.Error("QBFT: Failed to broadcast message", "msg", preprepare, "err", err)
+			return
+		}
+
+		// Set the preprepareSent to the current round
+		c.current.preprepareSent = curView.Round
+	}
+}
+
+func (c *core) handlePreprepareMsg(preprepare *message.Preprepare) error {
+	logger := c.logger.New("state", c.state)
+
+	c.logger.Info("QBFT: handlePreprepareMsg", "view", preprepare.View(), "m", preprepare)
+
+	// Check if the message comes from current proposer
+	logger.Warn("QBFT who's proposer?", "source", preprepare.Source(), "proposer", c.valSet.GetProposer().Address())
+	if !c.valSet.IsProposer(preprepare.Source()) {
+		logger.Warn("Ignore preprepare messages from non-proposer")
+		return errNotFromProposer
+	}
+
+	// Justification
+	if preprepare.Round.Uint64() > 0 && !justify(preprepare.Proposal, preprepare.JustificationRoundChanges, preprepare.JustificationPrepares, c.QuorumSize()) {
+		logger.Error("QBFT: Unable to justify PRE-PREPARE message")
```

Comment:
Should the message better be something like "QBFT: invalid PRE-PREPARE message's justification"?

Reply:
Also, should it be a warning (same as for invalid proposer)